src/meta/app/src/schema/index.rs

```diff
@@ -0,0 +1,303 @@
+
+    pub fn index_name(&self) -> String {
+        self.index_name.clone()
+    }
+
+    pub fn index_name_ident(&self) -> IndexNameIdent {
+        IndexNameIdent {
+            tenant: self.tenant.clone(),
+            index_name: self.index_name.clone(),
+        }
+    }
+}
+
+impl Display for IndexNameIdent {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "'{}'.'{}'", self.tenant, self.index_name)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq, Default)]
+pub struct IndexIdToName {
+    pub index_id: u64,
+}
+
+impl Display for IndexIdToName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.index_id)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct IndexId {
+    pub index_id: u64,
+}
+
+impl Display for IndexId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.index_id)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct IndexIdListKey {
+    pub tenant: String,
+    pub table_id: MetaId,
+}
+
+impl Display for IndexIdListKey {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "'{}'", self.tenant)
+    }
+}
+
+/// Save index id list history.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, Default, PartialEq)]
+pub struct IndexIdList {
+    pub id_list: Vec<u64>,
+}
+
+impl IndexIdList {
+    pub fn new() -> IndexIdList {
+        IndexIdList::default()
+    }
+
+    pub fn len(&self) -> usize {
+        self.id_list.len()
+    }
+
+    pub fn id_list(&self) -> &Vec<u64> {
+        &self.id_list
+    }
+
+    pub fn append(&mut self, index_id: u64) {
+        self.id_list.push(index_id);
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.id_list.is_empty()
+    }
+
+    pub fn pop(&mut self) -> Option<u64> {
+        self.id_list.pop()
+    }
+
+    pub fn last(&mut self) -> Option<&u64> {
+        self.id_list.last()
+    }
+}
+
+impl Display for IndexIdList {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "Index id list: {:?}", self.id_list)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+pub struct IndexMeta {
+    pub table_id: MetaId,
+    pub created_on: DateTime<Utc>,
+
+    // if used in CreateIndexReq, this field MUST set to None.
+    pub drop_on: Option<DateTime<Utc>>,
+    pub query: String,
+}
+
+impl Default for IndexMeta {
+    fn default() -> Self {
+        IndexMeta {
+            table_id: 0,
+            created_on: Utc::now(),
+            drop_on: None,
+            query: "".to_string(),
+        }
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct CreateIndexReq {
+    pub if_not_exists: bool,
+    pub name_ident: IndexNameIdent,
+    pub meta: IndexMeta,
+}
```

Comment:
I suggest moving `if_not_exists` out of this struct and let the caller determine what to do.

`if_not_exists` only affects the result returned to the user, not the behavior of the meta-service operation. Therefore, it would be better if `SchemaApi` returns either a **Created** or an **Exist** status code, and lets the caller decide whether to emit an error or not.

Reply:
But `CreateTableReq` also has this field, it's inconsistent to move it out.

Reply:
I have been thinking about removing the `if_not_exists` from the `CreateTableReq` as well. :)

Reply:
related issue: #11456